presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java

```diff
@@ -0,0 +1,117 @@
+import java.util.Optional;
+import java.util.Random;
+
+import static com.fasterxml.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toMap;
+
+public class ClusterManager
+{
+    private static final Random RANDOM = new Random();
+    private static final JsonCodec<RouterSpec> CODEC = new JsonCodecFactory(
+            () -> new JsonObjectMapperProvider().get().enable(FAIL_ON_UNKNOWN_PROPERTIES))
+            .jsonCodec(RouterSpec.class);
+
+    private final Map<String, GroupSpec> groups;
+    private final List<SelectorRuleSpec> groupSelectors;
+
+    @Inject
+    public ClusterManager(RouterConfig config)
+    {
+        RouterSpec routerSpec;
+        try {
+            routerSpec = CODEC.fromJson(Files.readAllBytes(Paths.get(config.getConfigFile())));
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        catch (IllegalArgumentException e) {
+            Throwable cause = e.getCause();
+            if (cause instanceof UnrecognizedPropertyException) {
+                UnrecognizedPropertyException ex = (UnrecognizedPropertyException) cause;
+                String message = format("Unknown property at line %s:%s: %s",
+                        ex.getLocation().getLineNr(),
+                        ex.getLocation().getColumnNr(),
+                        ex.getPropertyName());
+                throw new IllegalArgumentException(message, e);
+            }
+            if (cause instanceof JsonMappingException) {
+                // remove the extra "through reference chain" message
+                if (cause.getCause() != null) {
+                    cause = cause.getCause();
+                }
+                throw new IllegalArgumentException(cause.getMessage(), e);
+            }
+            throw e;
+        }
+
+        this.groups = ImmutableMap.copyOf(routerSpec.getGroups().stream().collect(toMap(GroupSpec::getName, group -> group)));
+        this.groupSelectors = ImmutableList.copyOf(routerSpec.getSelectors());
+    }
+
+    public List<URI> getAllClusters()
+    {
+        return groups.values().stream()
+                .flatMap(groupSpec -> groupSpec.getMembers().stream())
+                .collect(toImmutableList());
+    }
+
+    public Optional<URI> getDestination(RequestInfo requestInfo)
+    {
+        Optional<String> target = matchGroup(requestInfo);
+        if (!target.isPresent()) {
+            return Optional.empty();
+        }
+
+        checkArgument(groups.containsKey(target.get()));
+        List<URI> candidates = groups.get(target.get()).getMembers();
+        return Optional.ofNullable(candidates.get(RANDOM.nextInt(candidates.size())));
+    }
+
+    public Optional<String> matchGroup(RequestInfo requestInfo)
```

Comment:
Does this need to be public?

Reply:
Thanks for pointing that out! I updated it to `private`.